extensions/amp-analytics/0.1/amp-analytics.js

```diff
@@ -694,4 +719,100 @@ export class AmpAnalytics extends AMP.BaseElement {
   }
 }
 
+/**
+ * Keeps track of the third-party vendors' cross-domain iframes. Only one
+ * static instance will be created, which will keep manage ALL amp-analytics
+ * tags.
+ */
+class AmpAnalyticsFrames {
+  constructor() {
+    this.map = new Map();
+  }
+
+  /**
+   * Create a cross-domain iframe
+   * @param {!Element} parent  The document node of the parent page
+   * @param {!string} requestsFrameUrl  The URL, including data hash if
+   * applicable, of the cross-domain iframe
+   * @return {!Element}
+   */
+  create(parent, requestsFrameUrl) {
+    const context = {
+      "scriptSrc": "/examples/analytics-3p-remote-frame-helper.js",
+    };
+    let frame = createElementWithAttributes(parent, 'iframe', {
+      sandbox: 'allow-scripts',
+      name: JSON.stringify(context)
+    });
+    frame.onload = () => {
+      this.setIsReady(requestsFrameUrl);
+    };
+    frame.src = requestsFrameUrl; // Set AFTER onload to avoid race
+    frame.style.cssText='width:0;height:0;visibility:hidden;';
+    this.map.set(requestsFrameUrl, {
+      frame: frame,
+      isReady: false,
+      msgQueue: [],
+    });
+    return frame;
+  }
+
+  /**
+   * Returns whether a requestsFrameUrl is already known
+   * @param {!string} requestsFrameUrl
+   * @return {!boolean}
+   */
+  has(requestsFrameUrl) {
+    return this.map.has(requestsFrameUrl);
+  }
+
+  /**
+   * Sends a message to a cross-domain frame, or queues the message if the
+   * frame is not yet ready to receive messages.
+   * TODO: Implement throttling if messages are sent too rapidly.
+   * @param {!string} requestsFrameUrl
+   * @param {!string} message
+   */
+  send(requestsFrameUrl, message) {
+    try {
+      const mapEntry = this.map.get(requestsFrameUrl);
+      if (mapEntry.isReady) {
+        this.send_(mapEntry.frame, [message]);
+      } else {
+        mapEntry.msgQueue.push(message);
+      }
+    } catch (err) {
+      console.error("Failed to send event '" + err + "' to URL '" +
+        requestsFrameUrl + "'");
+    }
+  }
+
+  /**
+   * Send an array of messages to a cross-domain iframe
+   * @param {!Element} frame  The cross-domain iframe
+   * @param {!Array<string>} messages  The messages to send
+   * @private
+   */
+  send_(frame, messages) {
+    // Warning: the following code is likely only temporary. Don't check
+    // in before getting resolution on that.
+    frame &&
```

Comment:
should there also be an `&& frame.contentWindow` check here before you call a method on it?